cpp/src/precompile/gandiva.h

```diff
@@ -65,6 +55,58 @@ arrow::Decimal128 castDECIMAL(arrow::Decimal128 in, int32_t original_precision,
   return arrow::Decimal128(out);
 }
 
+arrow::Decimal128 castDECIMALNullOnOverflow(arrow::Decimal128 in, 
+                                            int32_t original_precision,
+                                            int32_t original_scale, 
+                                            int32_t new_precision,
+                                            int32_t new_scale,
+                                            bool* overflow_) {
+  bool overflow = false;
+  gandiva::BasicDecimalScalar128 val(in, original_precision, original_scale);
+  auto out = gandiva::decimalops::Convert(val, new_precision, new_scale, &overflow);
+  if (overflow) {
+    *overflow_ = true;
+  }
+  return arrow::Decimal128(out);
+}
+
+arrow::Decimal128 add(arrow::Decimal128 left, int32_t left_precision,
+                      int32_t left_scale, arrow::Decimal128 right,
+                      int32_t right_precision, int32_t right_scale,
+                      int32_t out_precision, int32_t out_scale) {
+  gandiva::BasicDecimalScalar128 x(left, left_precision, left_scale);
+  gandiva::BasicDecimalScalar128 y(right, right_precision, right_scale);
+  arrow::BasicDecimal128 out =
+      gandiva::decimalops::Add(x, y, out_precision, out_scale);
+  return arrow::Decimal128(out);
+}
+
+arrow::Decimal128 subtract(arrow::Decimal128 left, int32_t left_precision,
+                           int32_t left_scale, arrow::Decimal128 right,
+                           int32_t right_precision, int32_t right_scale,
+                           int32_t out_precision, int32_t out_scale) {
+  gandiva::BasicDecimalScalar128 x(left, left_precision, left_scale);
+  gandiva::BasicDecimalScalar128 y(right, right_precision, right_scale);
+  arrow::BasicDecimal128 out =
+      gandiva::decimalops::Subtract(x, y, out_precision, out_scale);
+  return arrow::Decimal128(out);
+}
+
+arrow::Decimal128 multiply(arrow::Decimal128 left, int32_t left_precision,
+                           int32_t left_scale, arrow::Decimal128 right,
+                           int32_t right_precision, int32_t right_scale,
+                           int32_t out_precision, int32_t out_scale) {
+  gandiva::BasicDecimalScalar128 x(left, left_precision, left_scale);
+  gandiva::BasicDecimalScalar128 y(right, right_precision, right_scale);
+  bool overflow = false;
+  arrow::BasicDecimal128 out =
+      gandiva::decimalops::Multiply(x, y, out_precision, out_scale, &overflow);
+  if (overflow) {
+    throw std::overflow_error("Decimal multiply overflowed!");
```

Comment:
will simply throw an exception raise some fail issue? shall we handle this overflow same as cast?